design/one-pager-considerations-for-changing-test-framework-of-uptest.md

```diff
@@ -0,0 +1,167 @@
+test Kubernetes Operators. `kuttl` is designed for testing operators, however it
+can declaratively test any kubernetes objects.
+
+While starting to the `Uptest` effort, we considered a few different
+alternatives and `kuttl`'s capabilities were appropriate for our assertion
+aims although its missing points. Today, we consider changing the underlying
+test framework tool because of the `kuttl` stopped to be actively maintained.
+
+## Goals
+
+- Determining the requirements of Uptest well and revealing its needs.
+- Having a more inclusive and capable underlying test framework.
+- Use of a more up-to-date and maintained tool in the background.
+
+## Proposal - Switching to `chainsaw` tool
+
+[chainsaw](https://github.com/kyverno/chainsaw) provides a declarative approach
+to test Kubernetes operators and controllers. While Chainsaw is designed for
+testing operators and controllers, it can declaratively test any Kubernetes
+objects. Chainsaw is an open-source tool that was initially developed for
+defining and running Kyverno end-to-end tests. The tool has Apache-2.0 license.
+
+In addition to the functionality provided by `kuttl`, the prominent features of
+the `chainsaw` tool are that it is maintained up to date and has an
+infrastructure for migration from `kuttl`.
+
+There are very similar concepts in `chainsaw` with `kuttl`. By this way, we do
+not have to make important changes in the templates.
+
+```yaml
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: example
+spec:
+  steps:
+  - try:
+    # ...
+    - apply:
+        file: my-configmap.yaml
+    # ...
+```
+
+```yaml
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: example
+spec:
+  steps:
+  - try:
+    # ...
+    - command:
+        entrypoint: echo
+        args:
+        - hello chainsaw
+    # ...
+```
+
+Also provides logical assertion statements:
+
+```yaml
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: example
+spec:
+  steps:
+  - try:
+    # ...
+    - assert:
+        resource:
+          apiVersion: v1
+          kind: Deployment
+          metadata:
+            name: foo
+          spec:
+            (replicas > 3): true
+    # ...
+```
+
+Resource Template support is another important requirement for Uptest:
+
+```yaml
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: template
+spec:
+  template: true
+  steps:
+  - assert:
+      resource:
+        # apiVersion, kind, name, namespace and labels are considered for templating
+        apiVersion: v1
+        kind: ConfigMap
+        metadata:
+          name: ($namespace)
+        # other fields are not (they are part of the assertion tree)
+        data:
+          foo: ($namespace)
+```
+
+Related Issue: https://github.com/upbound/official-providers-ci/issues/179
+
+In short, the `chainsaw` is more capable and maintained tool than `kuttl`. So,
+switching to this new tool may better for handling the Uptest requirements.
```

Comment:
```suggestion
In short, `chainsaw` is a more capable and well-maintained framework than `kuttl` and switching to it will better suit Uptest's future requirements.
```